src/algorithms/es512.rs

```diff
@@ -0,0 +1,416 @@
+}
+
+impl AsRef<P521VerifyingKey> for P521PublicKey {
+    fn as_ref(&self) -> &P521VerifyingKey {
+        &self.0
+    }
+}
+
+impl P521PublicKey {
+    pub fn from_bytes(raw: &[u8]) -> Result<Self, Error> {
+        let p521_pk =
+            P521VerifyingKey::from_sec1_bytes(raw).map_err(|_| JWTError::InvalidPublicKey)?;
+        Ok(P521PublicKey(p521_pk))
+    }
+
+    pub fn from_der(der: &[u8]) -> Result<Self, Error> {
+        let p521_pk =
+            P521VerifyingKey::from_public_key_der(der).map_err(|_| JWTError::InvalidPublicKey)?;
+
+        Ok(P521PublicKey(p521_pk))
+    }
+
+    pub fn from_pem(pem: &str) -> Result<Self, Error> {
+        let p521_pk =
+            P521VerifyingKey::from_public_key_pem(pem).map_err(|_| JWTError::InvalidPublicKey)?;
+        Ok(P521PublicKey(p521_pk))
+    }
+
+    pub fn to_bytes(&self) -> Vec<u8> {
+        self.0.to_encoded_point(true).as_bytes().to_vec()
+    }
+
+    pub fn to_bytes_uncompressed(&self) -> Vec<u8> {
+        self.0.to_encoded_point(false).as_bytes().to_vec()
+    }
+
+    pub fn to_der(&self) -> Result<Vec<u8>, Error> {
+        Ok(self
+            .0
+            .to_public_key_der()
+            .map_err(|_| JWTError::InvalidPublicKey)?
+            .as_ref()
+            .to_vec())
+    }
+
+    pub fn to_pem(&self) -> Result<String, Error> {
+        let p521_pk = p521::PublicKey::from(self.0);
+        Ok(p521_pk
+            .to_public_key_pem(Default::default())
+            .map_err(|_| JWTError::InvalidPublicKey)?)
+    }
+}
+
+#[doc(hidden)]
+pub struct P521KeyPair {
+    p521_sk: P521SigningKey,
+    metadata: Option<KeyMetadata>,
+}
+
+impl AsRef<P521SigningKey> for P521KeyPair {
+    fn as_ref(&self) -> &P521SigningKey {
+        &self.p521_sk
+    }
+}
+
+impl P521KeyPair {
+    pub fn from_bytes(raw: &[u8]) -> Result<Self, Error> {
+        let p521_sk = P521SigningKey::from_slice(raw).map_err(|_| JWTError::InvalidKeyPair)?;
+        Ok(P521KeyPair {
+            p521_sk,
+            metadata: None,
+        })
+    }
+
+    pub fn from_der(der: &[u8]) -> Result<Self, Error> {
+        let p521_sk = P521SigningKey::from_pkcs8_der(der).map_err(|_| JWTError::InvalidKeyPair)?;
+        Ok(P521KeyPair {
+            p521_sk,
+            metadata: None,
+        })
+    }
+
+    pub fn from_pem(pem: &str) -> Result<Self, Error> {
+        let p521_sk = P521SigningKey::from_pkcs8_pem(pem).map_err(|_| JWTError::InvalidKeyPair)?;
+        Ok(P521KeyPair {
+            p521_sk,
+            metadata: None,
+        })
+    }
+
+    pub fn to_bytes(&self) -> Vec<u8> {
+        self.p521_sk.to_bytes().to_vec()
+    }
+
+    pub fn to_der(&self) -> Result<Vec<u8>, Error> {
+        let scalar = NonZeroScalar::from_repr(self.p521_sk.to_bytes());
+        if bool::from(scalar.is_none()) {
+            return Err(JWTError::InvalidKeyPair.into());
+        }
+        let p521_sk =
+            P521SigningKey::from(NonZeroScalar::from_repr(scalar.unwrap().into()).unwrap());
```

Comment:
Is this unwrap safe or is it a leftover?

Reply:
It seems to be safe - it's the same code as ES256/384.